ugs-platform/ugs-platform-ugscore/src/main/java/com/willwinder/ugs/nbp/core/toolbars/Bundle.properties

```diff
@@ -0,0 +1,4 @@
+mainWindow.swing.firmware.toolbarTitle = Firmware selection toolbar
```

Comment:
An alternative for the MacrosService could be to create an action that presents a dynamic menu. Something like this:

```java
package com.willwinder.ugs.nbp.core.actions;

import com.willwinder.ugs.nbp.lib.lookup.CentralLookup;
import com.willwinder.universalgcodesender.MacroHelper;
import com.willwinder.universalgcodesender.model.BackendAPI;
import com.willwinder.universalgcodesender.types.Macro;
import org.openide.awt.*;
import org.openide.util.Exceptions;
import org.openide.util.actions.Presenter;

import javax.swing.*;
import java.awt.event.ActionEvent;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

@ActionID(
        id = "com.willwinder.ugs.nbp.core.actions.MacrosListAction",
        category = "Machine")
@ActionRegistration(
        lazy = false,
        displayName = "MacrosListAction")
@ActionReferences({
        @ActionReference(
                path = "Menu/Machine/Macros",
                position = 1000)
})
public class MacrosListAction extends AbstractAction implements DynamicMenuContent, Presenter.Popup {

    private final BackendAPI backend;

    public MacrosListAction() {
        this.backend = CentralLookup.getDefault().lookup(BackendAPI.class);
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        // does nothing, this is a popup menu
    }

    @Override
    public JComponent[] getMenuPresenters() {
        return createMenu();
    }

    @Override
    public JComponent[] synchMenuPresenters(JComponent[] items) {
        return createMenu();
    }

    @Override
    public JMenuItem getPopupPresenter() {
        JMenu menu = new JMenu(this);
        JComponent[] menuItems = createMenu();
        for (JComponent item : menuItems) {
            menu.add(item);
        }
        return menu;
    }

    @Override
    public boolean isEnabled() {
        return backend.isConnected() && backend.isIdle();
    }

    private JComponent[] createMenu() {
        Collection<Macro> macros = backend.getSettings().getMacros();
        List<JComponent> actions = macros.stream()
                .map(this::makeMenuItem)
                .collect(Collectors.toList());

        return actions.toArray(new JComponent[actions.size()]);
    }

    private JComponent makeMenuItem(Macro macro) {
        JMenuItem item = new JMenuItem();
        Actions.connect(item, new MacroAction(backend, macro), false);
        return item;
    }

    private class MacroAction extends AbstractAction {
        private BackendAPI backend;
        private Macro macro;

        public MacroAction(BackendAPI backend, Macro macro) {
            this.backend = backend;
            this.macro = macro;
            putValue(NAME, macro.getName());
            putValue("menuText", macro.getName());
            putValue("displayName", macro.getName());
        }

        @Override
        public void actionPerformed(ActionEvent e) {
            try {
                if (macro != null && macro.getGcode() != null) {
                    MacroHelper.executeCustomGcode(macro.getGcode(), backend);
                }
            } catch (Exception ex) {
                Exceptions.printStackTrace(ex);
            }
        }

        @Override
        public boolean isEnabled() {
            return backend.isConnected() && backend.isIdle();
        }
    }
}
```

This should also utilize the menu translations if it's defined in the layers.xml.